File: 20removeNode.js


// iterative
// n = number of nodes
// Time: O(n)
// Space: O(1)

const removeNode = (head, targetVal) => {
  if (head.val === targetVal) return head.next;
  
  let current = head;
  let prev = null;
  while (current !== null) {
    if (current.val === targetVal) {
      prev.next = current.next;
      break;
    }
    prev = current;
    current = current.next;
  }
  
  return head;
};

// recursive
// n = number of nodes
// Time: O(n)
// Space: O(n)

const removeNode = (head, targetVal) => {
  if (head === null) return null;
  if (head.val === targetVal) return head.next;
  head.next = removeNode(head.next, targetVal);
  return head;
};


// ----------------------------- //
// author: miguel
const removeNode = (head, targetVal) => {
  if (head.val === targetVal) return head.next;
  let slow = head;
  let fast = head.next;
  
  while (fast) {
    if (fast.val === targetVal) {
      slow.next = fast.next;
    }
    
    slow = slow.next;
    fast = fast.next;
  }
  return head;
};

File: Python/20removeNode.py

# recursive
# n = number of nodes
# Time: O(n)
# Space: O(n)
def remove_node(head, target_val):
  if not head:
    return None

  if head.val == target_val:
    return head.next

  head.next = remove_node(head.next, target_val)
  return head


# iterative
# n = number of nodes
# Time: O(n)
# Space: O(1)
def remove_node(head, target_val):
  if head.val == target_val:
    return head.next

  current = head
  prev = None

  while current:
    if current.val == target_val:
      prev.next = current.next
      break
    prev = current
    current = current.next

  return head
